account/validator.go

```diff
@@ -1,93 +1,187 @@
+// MIT License
+//
+// Copyright (c) 2018 Finterra
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
 package account
 
 import (
+	"encoding/hex"
 	"encoding/json"
+	"fmt"
+	"strconv"
 )
 
 type Validator interface {
 	Addressable
-	// The validator's voting power
-	Power() uint64
-	// Alter the validator's voting power by amount that can be negative or positive.
-	// A power of 0 effectively unbonds the validator
-	WithNewPower(uint64) Validator
-}
 
-// Neither abci_types or tm_types has quite the representation we want
-type ConcreteValidator struct {
-	Address   Address
-	PublicKey PublicKey
-	Power     uint64
+	Power() int64
+	Stake() uint64
+	Sequence() uint64
+	BondingHeight() uint64
+
+	MinimumStakeToUnbond() uint64
+
+	Bytes() ([]byte, error)
+	String() string
 }
 
-type concreteValidatorWrapper struct {
-	*ConcreteValidator `json:"unwrap"`
+type MutableValidator interface {
+	Validator
+
+	AddStake(stake uint64)
+	SubtractStake(stake uint64)
+	IncSequence()
 }
 
-var _ Validator = concreteValidatorWrapper{}
+type validator struct {
+	publicKey     PublicKey
+	stake         uint64
+	bondingHeight uint64
+	sequence      uint64
+}
 
-func AsValidator(account Account) Validator {
-	return ConcreteValidator{
-		Address:   account.Address(),
-		PublicKey: account.PublicKey(),
-		Power:     account.Balance(),
-	}.Validator()
+func NewValidator(publicKey PublicKey, stake, bondingHeight uint64) Validator {
+	return validator{
+		publicKey:     publicKey,
+		stake:         stake,
+		bondingHeight: bondingHeight,
+		sequence:      0,
+	}
 }
 
-func AsConcreteValidator(validator Validator) *ConcreteValidator {
-	if validator == nil {
+func LoadValidator(bytes []byte) Validator {
+	u := map[string]string{}
```

Comment:
I'm not that keen on this stringly typed way of encoding/decoding. Let's just define a struct that is a model of serialisation and call `json.Marshal` on that. Often we can use the same model of serialisation as model we use in-memory but if we can't I'd say just define one up front that clearly does the same thing on serialisation and deserialisation

Reply:
Having said that the approach with a wrapper (providing the interface) and a concrete type providing the serialisation has it's downsides so I am open to alternatives

Reply:
I am not that keen to define a structure just only for serializing/serializing. You know the `go` is a naive programming language and it's not written to make life easier! What we have is to choose between two pains!

Reply:
Strongly disagree on this - I have had to maintain hand spun serialisation logic like this and it is a pain. It's too easy to change a string - and crucially map order is *guaranteed* to be non-deterministic which in general we do not want if we are ever hashing things.

It is much better to have a single declarative struct that defines how a stored object can be serialiased *and* deserialised in a single location - exported so that other packages are able to handle that serialised object should they need to. Not to mention the fact you can attach documentation to the struct.

Interspersing code with the busywork of serialising integers is a mess. Defining a struct 'just' for serialisation is a great use of a struct. Once again this is also a case of folliowing existing conventions in the codebase if you still don't find the above convincing.

Reply:
Serializing can be done without relying on Go functionality. We can make deterministic serilizable function, (it's our own implementation).